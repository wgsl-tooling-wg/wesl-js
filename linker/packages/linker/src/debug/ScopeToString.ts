import { LineWrapper } from "./LineWrapper.ts";
import { Ident, Scope } from "../Scope.ts";

/** A debugging print of the scope tree with identifiers in nested brackets */
export function scopeToString(scope: Scope, indent = 0): string {
  const { children } = scope;
  let childStrings: string[] = [];
  if (children.length)
    childStrings = children.map(c => scopeToString(c, indent + 2));

  // list of identifiers, with decls prefixed with '%'
  const identStrings = scope.idents.map(({ kind, originalName }) => {
    const prefix = kind === "decl" ? "%" : "";
    return `${prefix}${originalName}`;
  });

  const str = new LineWrapper(indent);
  str.add("{ ");

  const last = identStrings.length - 1;
  identStrings.forEach((s, i) => {
    const element = i < last ? s + ", " : s;
    str.add(element);
  });

  if (childStrings.length) {
    str.nl();
    str.addBlock(childStrings.join("\n"));
  }

  if (str.oneLine) {
    str.add(" }");
  } else {
    if (!childStrings.length) str.nl();
    str.add("}");
  }

  return str.result;
}

export function identToString(ident?: Ident): string {
  if (!ident) return JSON.stringify(ident);
  const { kind, originalName } = ident;
  const idStr = ident.id ? `#${ident.id}` : "";
  if (kind === "ref") {
    const ref = identToString(ident.refersTo!);
    return `${originalName} ${idStr} -> ${ref}`;
  } else {
    return `%${originalName} ${idStr} (${ident.mangledName})`;
  }
}
