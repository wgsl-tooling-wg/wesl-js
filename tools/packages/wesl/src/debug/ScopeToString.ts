import { childScope, Ident, Scope } from "../Scope.ts";
import { attributeToString } from "./ASTtoString.ts";
import { LineWrapper } from "./LineWrapper.ts";

/** A debugging print of the scope tree with identifiers in nested brackets */
export function scopeToString(scope: Scope, indent = 0): string {
  const { contents, kind, ifAttributes = [] } = scope;

  const str = new LineWrapper(indent);
  const attrStrings = ifAttributes.map(a => attributeToString(a)).join(" ");
  if (attrStrings) str.add(attrStrings + " ");
  if (kind === "partial") str.add("-");
  str.add("{ ");

  const last = contents.length - 1;
  let lastWasScope = false;
  contents.forEach((elem, i) => {
    if (childScope(elem)) {
      const childScope: Scope = elem;
      const childBlock = scopeToString(childScope, indent + 2);
      !lastWasScope && str.nl();
      str.addBlock(childBlock);
      lastWasScope = true;
    } else {
      lastWasScope && str.add("  ");
      lastWasScope = false;
      const ident: Ident = elem;
      str.add(identShortString(ident));
      if (i < last) str.add(" ");
    }
  });

  if (str.oneLine) {
    str.add(" }");
  } else {
    str.add("}");
  }

  str.add(` #${scope.id}`);

  return str.result;
}

/** name of an identifier, with decls prefixed with '%' */
function identShortString(ident: Ident): string {
  const { kind, originalName } = ident;
  const prefix = kind === "decl" ? "%" : "";
  return `${prefix}${originalName}`;
}

export function identToString(ident?: Ident): string {
  if (!ident) return JSON.stringify(ident);
  const { kind, originalName } = ident;
  const idStr = ident.id ? `#${ident.id}` : "";
  if (kind === "ref") {
    const ref = identToString(ident.refersTo!);
    return `${originalName} ${idStr} -> ${ref}`;
  } else {
    const { mangledName } = ident;
    const mangled = mangledName ? `(${mangledName})` : "";
    return `%${originalName}${mangled} ${idStr} `;
  }
}
