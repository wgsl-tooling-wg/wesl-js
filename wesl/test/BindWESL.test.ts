import { expect, test } from "vitest";
import { scopeToStringLong } from "../debug/ScopeToString.ts";
import { bindTest } from "./TestUtil.ts";

test("nested scopes binding", () => {
  const src = `
    fn main() {
      let bar = 72;
      if (true) {
        if (true) {
          let new_bar = bar; // Should be 72!
        }
        let bar = 5;
      }
    }
  `;

  const { registry } = bindTest(src);
  const ast = registry.modules["package::test"];

  expect(scopeToStringLong(ast.rootScope)).toMatchInlineSnapshot();
});

test("@location attribute const", () => {
  const src = `
    const pos = 0;

    @fragment
    fn fragmentMain(@location(0) pos : vec3f) -> @location(pos) vec4f { 
      let x = pos;
    }
  `;
  const { registry } = bindTest(src);
  const ast = registry.modules["package::test"];
  expect(scopeToStringLong(ast.rootScope)).toMatchInlineSnapshot();
});
