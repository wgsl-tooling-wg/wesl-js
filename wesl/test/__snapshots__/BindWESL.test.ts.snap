export const snapshot = {};

snapshot[`nested scopes binding 1`] = `
"{ 
  -{ %main(main) #1  
    { %bar(bar) #3  
      { 
        { %new_bar #5  bar #7 -> %bar(bar) #3  } #5
        %bar #9 
      } #4
    } #2
  } #1
} #0"
`;

snapshot[`@location attribute const 1`] = `
"{ 
  -{ %pos(pos) #1  
    {  } #2
  } #1
  -{ %fragmentMain(fragmentMain) #3  pos #9 -> %pos(pos) #1 
     
    { %pos(pos) #5  vec3f #7 -> undefined 
      vec4f #11 -> undefined %x #13  pos #15 -> %pos(pos) #5 }
       #4
  } #3
} #0"
`;
